velox/functions/prestosql/window/FirstLastValue.cpp

```diff
@@ -82,9 +83,63 @@ class FirstLastValueFunction : public exec::WindowFunction {
     invalidRows_.applyToSelected([&](auto i) { rowNumbers_[i] = kNullRow; });
   }
 
+  void setRowNumbersRespectNulls(
+      const SelectivityVector& validRows,
+      const BufferPtr& frameStarts,
+      const BufferPtr& frameEnds) {
+    if constexpr (TValue == ValueType::kFirst) {
+      auto rawFrameStarts = frameStarts->as<vector_size_t>();
+      validRows.applyToSelected(
+          [&](auto i) { rowNumbers_[i] = rawFrameStarts[i]; });
+    } else {
+      auto rawFrameEnds = frameEnds->as<vector_size_t>();
+      validRows.applyToSelected(
+          [&](auto i) { rowNumbers_[i] = rawFrameEnds[i]; });
+    }
+  }
+
+  void setRowNumbersIgnoreNulls(
+      const SelectivityVector& validRows,
+      const BufferPtr& frameStarts,
+      const BufferPtr& frameEnds) {
+    auto extractNullsResult = partition_->extractNulls(
+        valueIndex_, validRows, frameStarts, frameEnds, &nulls_);
+
+    auto firstFrame = extractNullsResult.first;
+    auto frameSize = extractNullsResult.second;
+    if (bits::findFirstBit(nulls_->as<uint64_t>(), 0, frameSize) < 0) {
+      // There are no nulls in the column. Continue the processing with the
+      // function that respects nulls since it is more efficient.
+      return setRowNumbersRespectNulls(validRows, frameStarts, frameEnds);
+    }
+
+    // nulls_ now has the non-null bits set.
+    bits::negate(nulls_->asMutable<char>(), frameSize);
+    auto rawNulls = nulls_->as<uint64_t>();
+
+    auto rawFrameStarts = frameStarts->as<vector_size_t>();
+    auto rawFrameEnds = frameEnds->as<vector_size_t>();
+    validRows.applyToSelected([&](auto i) {
+      auto frameStart = rawFrameStarts[i];
+      auto frameEnd = rawFrameEnds[i];
+      // Find first or last non-null position.
```

Comment:
nit: would it help to add findFirstNullBit to Nulls.h ?

Reply:
Didn't follow the suggestion. I used the utility functions from BitUtil.h itself.

Reply:
This comment is pretty verbose and makes reading the code a bit hard. I wonder if it would be helpful to add a helper function to Nulls.h which hides the details about whether isNull is 1 or 0 and make it not necessary to add this comment making the code easier to read.

Reply:
I realize I've written the comments in a way that might be hard to read. Will change this to make the reading easier.